src/Database/Expression/CaseExpressionInterface.php

```diff
@@ -0,0 +1,332 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.3.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Database\Expression;
+
+use Cake\Database\ExpressionInterface;
+use Cake\Database\TypedResultInterface;
+use Cake\Database\TypeMap;
+
+interface CaseExpressionInterface extends ExpressionInterface, TypedResultInterface
+{
+    /**
+     * Returns the case value for a `CASE case_value WHEN ...` expression.
+     *
+     * @return \Cake\Database\ExpressionInterface|object|scalar|null
+     */
+    public function getValue();
+
+    /**
+     * Sets the value for the case expression.
+     *
+     * When a value is set, the syntax generated is
+     * `CASE case_value WHEN when_value ... END`, where the
+     * `when_value`'s are compared against the `case_value`.
+     *
+     * When no value is set, the syntax generated is
+     * `CASE WHEN when_conditions ... END`, where the conditions
+     * hold the comparisons.
+     *
+     * @param \Cake\Database\ExpressionInterface|object|scalar|null $value The case value.
+     * @param string|null $valueType The case value type.
+     * @return $this
+     */
+    public function value($value, ?string $valueType = null);
+
+    /**
+     * Returns the case value type.
+     *
+     * @return string|null
+     */
+    public function getValueType(): ?string;
+
+    /**
+     * Returns the `WHEN ... THEN ...` statement expressions.
+     *
+     * @return \Cake\Database\Expression\WhenThenExpressionInterface[]
+     */
+    public function getWhen(): array;
```

Comment:
Do you need all the getter methods? Could something like `Query::clause()` work as a general purpose accessor.

Reply:
They're not necessarily needed, no. I don't remember the exact use case that made me add them initially, that was quite some time ago. The type getters were introduced later, actually _needed_ was just `WhenThenExpression::getThenType()` so that the return type for `CaseStatementExpression` could be guessed, and then this lonely getter seemed weird, so I added some friends.

Something like `Query::clause()` could definitely work. Not sure what it could look like for types, maybe the types would be separate "clauses", or a clause would be an associative array that holds the value and the type... 🤔

Reply:
Couldn't the clauses of this be `when`, `then` and `else`?

Reply:
I don't really have a use case for any of the types other than the type for the nested `WhenThenExpression::then()`.

How about not exposing any types other than that one, it could be a method like `WhenThenExpression::getResultType()`, making things a bit more abstract/generic?